cdc/split.cc

```diff
@@ -319,6 +320,35 @@ set_of_changes extract_changes(const mutation& base_mutation) {
 
 namespace cdc {
 
+struct find_timestamp_visitor {
+    api::timestamp_type _ts = api::missing_timestamp;
+
+    inline bool finished() const { return _ts != api::missing_timestamp; }
+
+    inline void visit(api::timestamp_type ts) { _ts = ts; }
+    inline void visit(const atomic_cell_view& cell) { visit(cell.timestamp()); }
+
+    void live_atomic_cell(const column_definition&, const atomic_cell_view& cell) { visit(cell); }
+    void dead_atomic_cell(const column_definition&, const atomic_cell_view& cell) { visit(cell); }
+    void collection_tombstone(const tombstone& t) {
+        // A collection tombstone with timestamp T can be created with:
+        // UPDATE ks.t USING TIMESTAMP T + 1 SET X = null WHERE ...
+        // (where X is a collection column).
+        // This is, among others, the reason why we show it in the CDC log
+        // with cdc$time using timestamp T + 1 instead of T.
+        visit(t.timestamp + 1);
+    }
+    void live_collection_cell(bytes_view, const atomic_cell_view& cell) { visit(cell); }
+    void dead_collection_cell(bytes_view, const atomic_cell_view& cell) { visit(cell); }
+    void collection_column(const column_definition&, auto visit_collection) { visit_collection(*this); }
```

Comment:
I'm wondering whether `auto visit_collection` shouldn't be `auto&& visit_collection`?

Reply:
Yeah, I'm wondering that too.
If the compiler is stupid:
- `auto` could result in copying a lambda. The lambda holds a reference (see how `collection_column` is called in `inspect_row_cells`), so at least even if it happened, it wouldn't hurt us much.
- `auto&&` could result in an indirection when we call the lambda.
I dunno. I guess we can go with `auto&&` since it's the most common approach probably.